### ISSUE TITLE
Sensor Questions

### ISSUE THREAD

**Issue by shayes83** [magico13/ha-emporia-vue#200]
Thank you very much for creating this integration.  As I was reading some comments through this forum, I got to thinking about how the 1 MIN sensor values are calculated.  if it is the avg power over a 1 minute period, how does the sensor accurately calculate a load that is more transient such as a well pump that always has a large draw but is always less than a minute in duration?  

Secondly, would there be a way to pull real-time values other than flashing and manipulating the unit itself?

Again, very nice work!

**Comment by magico13** [comment]
The Emporia unit itself sends data up to the cloud at least once a second, if not more often, and then the 1 minute sensor data is compiled from that data when the integration asks for it. So for something that draws 1000 watts for 15 seconds and 0 watts for the other 45 seconds, their database will have the data at the seconds level, so when it gets the average you should get a response saying 250 watts over the minute.

To get more technical the API actually returns the energy used over the last minute, which is just the sum of the Voltage x Current x TimeSinceLastReading measurements that the unit sends to the cloud, then that amount is scaled to get the average power.

When the integration asks for the 1 day or 1 month levels, the API will sum up all of the individual readings that were sent to the cloud during that time period and return the total. Since that's already in kwh there's no scaling there. In theory the 1D values should be the same as summing up all the 1 minute values once a minute, which the integration actually does automatically to smooth out the 1D data between requests, but every now and then the API returns bad data for a 1 minute request so that's why the integration still requests the 1D data from the API every 15 minutes or so.

That's a bit long winded and probably more than you needed, but hopefully answers your first question.

As for closer to realtime data, sadly Emporia doesn't provide a way for this integration to get that without modifying the device. The creator of Home Assistant even reached out to them to try to convince them to provide a native local option but they don't seem interested and instead want users to use their cloud. There was a brief period of time where the integration provided data at the 1 second level but Emporia asked me to remove that because we were using a disproportionate amount of their processing. You can still write a custom python script to get at that data (anything the App can do can theoretically be done with the API) but they recommend getting that data infrequently and just requesting the historical data. Unfortunately as far as I know I can't put historical data into Home Assistant, it expects the current value at each update. There are other projects that do pull that historical data for you and graph it but they're outside of Home Assistant and aren't realtime either.

Basically, if you want realtime data in Home Assistant you have to reflash it with a custom firmware and lose the cloud functionality. If you want historical data at 1 second intervals you can get that with another tool using the API. If you just want an "easy" way to get mostly correct energy data into Home Assistant for the energy page and some automations then this integration is that.

**Comment by Kathou99** [comment]
@magico13 
So did is somethings I can make it work to pull the data for each sec? Like I will do a pull resquest to a version with the sec options? or you will do it? because Emporia use second

Thank :)

**Comment by magico13** [comment]
@Kathou99 Despite it being possible, it is not permitted by Emporia. The community has specifically been asked to keep the number of requests below 2000 a day so requesting data every minute is fine but any more frequent is problematic.
